lib/ExtraCorres.thy

```diff
@@ -251,6 +249,88 @@ lemma hoare_from_abs_inv:
+  assumes
+    "\<And>s. \<lbrace>\<lambda>s'. P s \<and> P' s' \<and> (s, s') \<in> srel\<rbrace>
+          f' \<lbrace>\<lambda>rv' t'. \<exists>(rv, t) \<in> fst (f s). (t, t') \<in> srel \<and> rrel rv rv'\<rbrace>"
+  shows "corres_underlying srel nf nf' rrel P P' f f'"
+  using assms
+  by (fastforce simp: corres_underlying_def valid_def no_fail_def)
+
+lemma corres_from_valid_det:
+  assumes det: "det_wp P f"
+  assumes nf': "nf' \<Longrightarrow> no_fail (P' and ex_abs_underlying srel P) f'"
+  assumes valid:
+    "\<And>s rv t.
+      \<lbrakk>fst (f s) = {(rv, t)}; P s\<rbrakk>
+      \<Longrightarrow> \<lbrace>\<lambda>s'. P' s' \<and> (s, s') \<in> srel\<rbrace> f' \<lbrace>\<lambda>rv' t'. (t, t') \<in> srel \<and> rrel rv rv'\<rbrace>"
+  shows "corres_underlying srel nf nf' rrel P P' f f'"
+  apply (clarsimp simp: corres_underlying_def)
+  apply (intro conjI)
+   apply (insert det)
+   apply (clarsimp simp: det_wp_def)
+   apply (force dest: use_valid[OF _ valid])
+  apply (fastforce dest: nf' simp: no_fail_def ex_abs_underlying_def)
+  done
+
+lemma corres_noop_ex_abs:
+  assumes P: "\<And>s. P s \<Longrightarrow> \<lbrace>\<lambda>s'. (s, s') \<in> sr \<and> P' s'\<rbrace> f \<lbrace>\<lambda>rv s'. (s, s') \<in> sr \<and> r x rv\<rbrace>"
+  assumes nf': "nf' \<Longrightarrow> no_fail (P' and ex_abs_underlying sr P) f"
+  shows "corres_underlying sr nf nf' r P P' (return x) f"
+  apply (simp add: corres_underlying_def return_def)
+  apply clarsimp
+  apply (frule P)
+  apply (insert nf')
+  apply (fastforce simp: valid_def no_fail_def ex_abs_underlying_def)
+  done
+
+lemma corres_symb_exec_r_conj_ex_abs:
+  assumes z: "\<And>rv. corres_underlying sr nf nf' r Q (R' rv) x (y rv)"
+  assumes y: "\<lbrace>Q'\<rbrace> m \<lbrace>R'\<rbrace>"
+  assumes x: "\<And>s. Q s \<Longrightarrow> \<lbrace>\<lambda>s'. (s, s') \<in> sr \<and> P' s'\<rbrace> m \<lbrace>\<lambda>rv s'. (s, s') \<in> sr\<rbrace>"
+  assumes nf: "nf' \<Longrightarrow> no_fail (P' and ex_abs_underlying sr Q) m"
+  shows      "corres_underlying sr nf nf' r Q (P' and Q') x (m >>= (\<lambda>rv. y rv))"
+proof -
+  have P: "corres_underlying sr nf nf' dc Q P' (return undefined) m"
+    apply (rule corres_noop_ex_abs)
+     apply (simp add: x)
+    apply (erule nf)
+    done
+  show ?thesis
+  apply (rule corres_guard_imp)
+    apply (subst return_bind[symmetric], rule corres_split[OF P])
+      apply (rule z)
+     apply wp
+    apply (rule y)
+   apply simp+
+  done
+qed
+
+lemmas corres_symb_exec_r_conj_ex_abs_forwards =
+  corres_symb_exec_r_conj_ex_abs[where P'=P' and Q'=P' for P', simplified]
+
+lemma gets_the_corres_ex_abs:
+ "\<lbrakk>no_ofail P a; no_ofail (P' and ex_abs_underlying sr P) b\<rbrakk> \<Longrightarrow>
+   corres_underlying sr False True r P P' (gets_the a) (gets_the b)
+   = (\<forall> s s'. P s \<and> P' s' \<and> (s, s') \<in> sr \<longrightarrow> r (the (a s)) (the (b s')))"
```

Comment:
Phrasing corres rules as equalities is usually dangerous, because `simp` doesn't play nice with the schematics we usually have in those goals. If it's important to have an equality in this one, I'd vote for providing an additional rule with `[THEN iffD2]` that can be used as an intro rule in the goal.

Reply:
It seems like `gets_the_corres` is used either with `subst`, or with a `[THEN iffD1/2]`. Are you wanting a `lemmas gets_the_corres'` or something that does these `[THEN iffD1/2]` things for us?

This is a generalisation of `gets_the_corres`, which we have on the `rt` branch, and which is one of the few rules which hasn't made its way over to `master`. Should I do anything with that rule, too? Should I put that into `master` too, as part of this PR?

Reply:
I would call this rule (with equality) `gets_the_corres_ex_abs'`, because it's not in standard corres rule form (which are basically all `intro` rules for `corres` with different patterns). And yes, my `gets_the_corres_ex_abs` would just be a `lemmas .. = gets_the_corres_ex_abs'[THEN iffD2]` (or iffD1, I can never remember the direction), so people don't have to do that themselves and don't get tempted to just chuck the equality into `simp`.

Bringing the `gets_the_corres` rule over is a good idea in any case, though. If this one here is more general, we should possibly derive the other one from it, unless the proofs are trivial.

Reply:
I've pushed one commit for these things. `iffD2` was the right one to make the `corres_underlying` appear in the conclusion that that they're `intro` rules. And I do use the other direction for another lemma (`gets_the_corres_relation` just below) so yes, we should keep the equality too.

The proof for these was a lot of unfolding of basic definitions, so I did use the more general one to proof the less general one.